### ISSUE TITLE
Traits and associated types are not properly resolved in trait clauses

### ISSUE THREAD

**Issue by nox** [rust-lang/rust#48674]
https://play.rust-lang.org/?gist=408db89ed821df7b75a2769ae92d3e9b&version=nightly

**Comment by nox** [comment]
@nikomatsakis Can this get mentored? Is this blocked on Chalk? Can I help with it? Can we get any kind of ETA?

**Comment by rkarp** [comment]
Looks like another duplicate of #24159.

**Comment by nikomatsakis** [comment]
@nox left a comment on #24159

**Comment by estebank** [comment]
We now suggest adding the appropriate bound:

```
error[E0277]: the trait bound `P: F<()>` is not satisfied
  --> src/main.rs:17:1
   |
17 | / impl<P> M for A<P> where
18 | |     N<P>: M,
19 | |     //P: F<()>,
20 | |     P: F<N<S>>, // This should work just as fine as the commented bound just above
   | |                - help: consider further restricting type parameter `P`: `, P: F<()>`
21 | | {}
   | |__^ the trait `F<()>` is not implemented for `P`

error[E0277]: the trait bound `P: F<()>` is not satisfied
  --> src/main.rs:17:9
   |
17 | impl<P> M for A<P> where
   |         ^ the trait `F<()>` is not implemented for `P`
...
20 |     P: F<N<S>>, // This should work just as fine as the commented bound just above
   |                - help: consider further restricting type parameter `P`: `, P: F<()>`
   |
   = note: required because of the requirements on the impl of `M` for `A<P>`
```

**Comment by estebank** [comment]
Triage: no change.
